File: TwoSumClosestPair.py
def closest_to_zero(arr):
    arr.sort() # Sắp xếp mảng theo thứ tự tăng dần
    i, j = 0, len(arr) - 1  # Khởi tạo chỉ số i (đầu mảng) và j (cuối mảng)
    # Khởi tạo tổng bằng tổng của phần tử đầu và phần tử cuối
    sum_val = arr[i] + arr[j]
    # Khởi tạo biến diff bằng giá trị tuyệt đối của tổng ban đầu
    diff = abs(sum_val)
    while i < j:  # Lặp qua mảng cho đến khi i gặp j
        # Nếu tổng của hai phần tử bằng 0, trả về 0 vì không có cặp nào tốt hơn
        if arr[i] + arr[j] == 0:
            return 0
        # Nếu tổng của cặp phần tử có giá trị tuyệt đối nhỏ hơn giá trị hiện tại
        # thì cập nhật lại kết quả
        if abs(arr[i] + arr[j]) < abs(diff):
            diff = arr[i] + arr[j]
            sum_val = arr[i] + arr[j]
        # Nếu có nhiều cặp có tổng tuyệt đối giống nhau, chọn cặp có tổng lớn hơn
        elif abs(arr[i] + arr[j]) == abs(diff):
            sum_val = max(sum_val, arr[i] + arr[j])
        # Nếu tổng lớn hơn 0, giảm chỉ số j để tìm tổng nhỏ hơn
        if arr[i] + arr[j] > 0:
            j -= 1
        # Nếu tổng nhỏ hơn 0, tăng chỉ số i để tìm tổng lớn hơn
        else:
            i += 1
    return sum_val  # Trả về tổng gần 0 nhất

arr = list(map(int, input("Nhập các số trong mảng, cách nhau bởi khoảng trắng: ").split()))
print(closest_to_zero(arr))  # In kết quả

File: TwoSumClosestPair.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
// Hàm so sánh để sử dụng với qsort
int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}
// Hàm tìm cặp có tổng gần 0 nhất
int closestToZero(int arr[], int n) {
    // Sắp xếp mảng theo thứ tự tăng dần
    qsort(arr, n, sizeof(int), compare);
    // Khởi tạo hai con trỏ, một ở đầu mảng, một ở cuối mảng
    int i = 0, j = n - 1;
    // Khởi tạo tổng ban đầu là tổng của phần tử đầu và cuối mảng
    int sum = arr[i] + arr[j];
    int diff = abs(sum);  // Khởi tạo hiệu số tuyệt đối ban đầu
    while (i < j) {
        // Nếu tổng bằng 0, trả về 0 ngay lập tức
        if (arr[i] + arr[j] == 0) {
            return 0;
        }
        // Tính tổng của cặp phần tử hiện tại và hiệu số tuyệt đối
        int current_sum = arr[i] + arr[j];
        int current_diff = abs(current_sum);
        // Nếu hiệu số tuyệt đối nhỏ hơn hiệu số trước đó, cập nhật kết quả
        if (current_diff < diff) {
            diff = current_diff;
            sum = current_sum;
        } else if (current_diff == diff) {
            // Nếu có nhiều cặp có hiệu số tuyệt đối bằng nhau, chọn cặp có tổng lớn hơn
            sum = (sum > current_sum) ? sum : current_sum;
        }
        // Di chuyển con trỏ dựa trên tổng hiện tại
        if (current_sum > 0) {
            j--;  // Nếu tổng dương, giảm con trỏ bên phải
        } else {
            i++;  // Nếu tổng âm, tăng con trỏ bên trái
        }
    }
    return sum;  // Trả về tổng gần 0 nhất
}

int main() {
    int n; printf("Nhập số phần tử của mảng: "); scanf("%d", &n);
    int arr[n]; printf("Nhập các phần tử của mảng:\n");
    for (int i = 0; i < n; i++) {
        printf("arr[%d] = ", i);
        scanf("%d", &arr[i]);
    }
    printf("Tổng gần 0 nhất là: %d\n", closestToZero(arr, n));
    return 0;
}
